### PR TITLE
Add free variable counts in the formals to Receive as a field

### PR DESCRIPTION
**Pull request by KentShikama**
This resolves https://rchain.atlassian.net/browse/RHOL-139

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 1]
```diff
@@ -193,16 +193,19 @@ object ProcNormalizeMatcher {
             (result.chan :: acc._1, result.knownFree)
           }
         )
-        val newEnv = input.env.absorbFree(formalsResults._2)
+        val newEnv       = input.env.absorbFree(formalsResults._2)
+        val newFreeCount = formalsResults._2.next - input.knownFree.next
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: It should just be formalsResults._2.next.
What you have may even end up being negative.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala [hunk 2]
```diff
@@ -64,7 +64,11 @@ case class Send(chan: Channel, data: List[Par], persistent: Boolean)
 // [Par] is an n-arity Pattern.
 // It's an error for free Variable to occur more than once in a pattern.
 // Don't currently support conditional receive
-case class Receive(binds: List[(List[Channel], Channel)], body: Par, persistent: Boolean)
+// Count is the number of free variables in the formals
+case class Receive(count: Int,
+                   binds: List[(List[Channel], Channel)],
+                   body: Par,
+                   persistent: Boolean)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: Please put it at the end.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SortTest.scala [hunk 3]
```diff
@@ -410,39 +410,44 @@ class ParSortMatcherSpec extends FlatSpec with Matchers {
     result.term should be (sortedParExpr)
   }
 
-  "Par" should "Sort Receives based on persistence and then channels and then patterns and then body" in {
+  "Par" should "Sort Receives based on count, persistence, channels, patterns and then body" in {
```
Comment (guardbotmk3) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/SortTest.scala]: You shouldn't need to sort on count. It's utility info.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala (hunk 1 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Yeah the formalsResults starts with its own empty map; thanks for the catch.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala (hunk 2 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala]: Done

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SortTest.scala (hunk 3 above)
Comment (KentShikama) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/SortTest.scala]: Done.

**Comment by KentShikama** [comment]
@guardbotmk3 Ok ready for another round of reviews.

**Comment by KentShikama** [comment]
Yeah I did that with commit 6623121.

**Pull request merged**